driver-core/src/main/java/com/datastax/driver/core/Metadata.java

```diff
@@ -1077,14 +1112,14 @@ private static TokenMap build(
           factory, ring, tokenRanges, tokenToPrimary, allTokens, tokenToHosts, hostsToRanges);
     }
 
-    private Set<Host> getReplicas(String keyspace, Token token) {
+    private List<Host> getReplicas(String keyspace, Token token) {
 
-      Map<Token, Set<Host>> tokenToHosts = tokenToHostsByKeyspace.get(keyspace);
-      if (tokenToHosts == null) return Collections.emptySet();
+      Map<Token, List<Host>> tokenToHosts = tokenToHostsByKeyspace.get(keyspace);
+      if (tokenToHosts == null) return Collections.emptyList();
 
       // If the token happens to be one of the "primary" tokens, get result directly
-      Set<Host> hosts = tokenToHosts.get(token);
-      if (hosts != null) return hosts;
+      List<Host> hosts = tokenToHosts.get(token);
+      if (hosts != null) return new ArrayList<>(hosts);
```

Comment:
Previous implementation did not instantiate a new object. Is this necessary?

Reply:
We need to make sure that List read from tokenToHosts is not written at any point.
There are two options - either use `Collections.unmodifiableList` or copy it.
Using `Collections.unmodifiableList` will require checking if List is writable before doing anything on it.
Which can only be done via `replicas.getClass().getName().contains("Unmodifiable")` and I don't like it.
If you think we better go that way, let's discuss it.

Reply:
> Using Collections.unmodifiableList will require checking if List is writable before doing anything on it.
This should not be a concern anyway, right? We don't want it to be written, so we won't need to check that.
In that case we can use `ImmutableList`.

Reply:
Thanks, that turned out was good idea, thansk a lot, take a look now please.